db/query.go

```diff
@@ -170,6 +171,28 @@ var QueryPrincipals = SGQuery{
 	adhoc: false,
 }
 
+var QueryPrincipalsExcludeDeleted = SGQuery{
+	name: QueryTypePrincipalsExcludeDeleted,
+	statement: fmt.Sprintf(
+		"SELECT META(%s).id "+
+			"FROM %s AS %s "+
+			"USE INDEX($idx) "+
+			"WHERE META(%s).id LIKE '%s' "+
+			"AND (META(%s).id LIKE '%s' "+
+			"OR META(%s).id LIKE '%s') "+
+			"AND (%s.deleted IS MISSING OR %s.deleted = false) "+
+			"AND META(%s).id >= $%s "+ // Uses >= for inclusive startKey
+			"ORDER BY META(%s).id",
+		base.KeyspaceQueryAlias,
+		base.KeyspaceQueryToken, base.KeyspaceQueryAlias,
+		base.KeyspaceQueryAlias, SyncDocWildcard,
+		base.KeyspaceQueryAlias, `\\_sync:user:%`,
```

Comment:
Given that our only use case is for role retrieval, and we're having to build a new query anyway, I'd suggest omitting users from this query.  Deployments with a very large number of users and a very small number of roles are common, and it would be nice to avoid the query fetch on each user, since this isn't a covering index.

Reply:
the usage below seems to require both, users and roles... but I guess I could revert and leave `AllPrincipalIDs` and `QueryPrincipals` as they are (retrieve all), and create new functions just for active roles, something like `GetActiveRoleIDs`, that is only called by the API ?

https://github.com/couchbase/sync_gateway/blob/b630ddf8be5714cb357aa79aed96dfa9678e726c/db/database.go#L1547

(and maybe here)
https://github.com/couchbase/sync_gateway/blob/b630ddf8be5714cb357aa79aed96dfa9678e726c/db/database.go#L1498